linkerd/proxy/http/src/overwrite_authority.rs

```diff
@@ -0,0 +1,157 @@
+use super::h1;
+use futures::{try_ready, Future, Poll};
+use http;
+use http::header::HOST;
+use http::header::{HeaderValue, IntoHeaderName};
+use http::uri::Authority;
+use tracing::trace;
+
+pub trait ExtractAuthority<T> {
+    fn extract(&self, target: &T) -> Option<Authority>;
+}
+
+pub trait ForceAbsForm {
+    fn is_abs_form(&self) -> bool;
+}
+
+#[derive(Debug, Clone)]
+pub struct Layer<E, H> {
+    extractor: E,
+    canonical_dst_header: H,
+}
+
+#[derive(Clone, Debug)]
+pub struct MakeSvc<E, H, M> {
+    extractor: E,
+    canonical_dst_header: H,
```

Comment:
It's surprising to me that this is coupled to the canonical dst header. Why?

Reply:
So.. yes in theory we can do that. It just made sense to me that we have this separate layer that overwrites `authoirty` and the canonical_dst header as a separate step, even though the `canonical_dst` was already set. Also, seems to me the `canonical_dst` is set from the `Target.Addr` at the moment via `header_from_target`, which is not modified by the authority_overwrite layer.  We can split all of that up.  I am not sure coupling the authority overwrite to the canonical_dst  troubles me so much as we actually want to overwrite both, right?

Reply:
My view is that this module shouldn't know anything about canonical dst headers. What other headers does it need to know about? This shouldn't have to grow to know about arbitrary Linkerd headers. Should this module take a list of arbitrary headers to overwrite? How do we decouple this from this particular linkerd feature.

Reply:
Is it even appropriate to set the canonical dst header, since we're setting a canonical authority anyway? Is it more appropriate to clear it entirely? How do we expect service profiles to interact with this in general?

Reply:
@olix0r Good questions! So the reason I am overwriting the canonical dst and host at that point is because of the situation that arises when doing traffic splitting. What happens is the following: 

Lets say we have a ts that looks like: 

apex-svc: backend-zero-svc (on the local cluster)
leafsvc: backend-two-svc-remote (on the remote cluster). 

If we do not set the canonical dst the request will leave the proxy with the floowing values for these headers: 

```
l5d-dst-canonical: backend-zero-svc.default.svc.cluster.local:8888
host: backend-zero-svc:8888
```

So when this request arrives at the remote gateway, it cannot be routed. This is why I decided to overwrite these at that point. Maybe I am missing the context as to whether this is appropriate or not, but it seemed correct to me. Wrt to service profiles, I am not sure thie interferes with them too much. The profiles on the outbound side a resolved before this rewriting is applied. The profiles on the incoming side (the remote cluster) are resolved based on the canonical_dst that we set, which will be equal to the name of the service that actually exists on the remote cluster. So this seems correct to me. I guess the same thing will work if we blank the value of this header, because then the incoming Recognize will rely on the authority of the request.  So any suggestion as to how we want to handle that without coupling are more than welcome :)

Reply:
Okay, I follow what you're saying. Here's what's tripping me up: within a cluster, service profiles are only ever applied on a logical name (i.e. in the face of a traffic split, the inbound proxy uses the canonical-form logical dst to resolve the service profile). Typically, both the outbound and inbound proxies refer to the same service profile data (though there's some added complexity about why that's not always the case). So, I guess my question is: as a user, what sort of service name would i reference when configuring a service profile for inbound traffic on the remote cluster? Do we have a concrete example of the routing logic in play for a traffic-split that goes through a gateway? That would help me be clearer on this.

My general concern about the coupling with canonical-dst is that if there's one such case, there are likely to be others. Does it make sense to take a list of headers to clear in this case? Or should we support adding a `Proxy` (as we do for service profile routes) to support adding arbitrary logic to gateway traffic?